discretize/tree_mesh.py

```diff
@@ -359,6 +360,316 @@ def origin(self, value):
         # then update the TreeMesh with the hidden value
         self._set_origin(self._origin)
 
+    def refine_bounding_box(self, points, level, padding_cells_by_level=None, finalize=True, diagonal_balance=None):
+        """Refine within a bounding box based on the maximum and minimum extent of scattered points.
+
+        This function refines the tree mesh based on the bounding box defined by the
+        maximum and minimum extent of the scattered input `points`. It will refine
+        the tree mesh for each level given.
+
+        It also optionally pads the bounding box at each level based on the number of
+        padding cells at each dimension.
+
+        Parameters
+        ----------
+        points : (N, dim) array_like
+            The bounding box will be the maximum and minimum extent of these points
+        level : (n_level) array_like of int
+            The level of the treemesh to refine the bounding box to. If an array it will
+            refine the box to each level (mostly useful in combination with `padding_cells_by_level`)
```

Comment:
I am not getting this part - Feels like the level as `array` only matters if you apply padding_cells. Should we not restrict `level: int` for this one?

Reply:
ok I get it now - you are trying not to impose order in the values of levels. Feels like we are increasing the complexity of input arguments for very little benefit. I would propose we revert to combining `level` and `padding_cells_by_level` into a single `list[int]`.

Reply:
I would rather leave the refinement level as an explicit input argument that is not inferred by the length of an input list.

Reply:
Let s do the dictionary. More obvious that the two arguments depend on each other. Otherwise there is weird hidden dependency IMO

Reply:
It seems like the basic concept is that:
- we have a single box
- everywhere within that box will be discretize to the finest level provided (which means *level* makes more sense as an *int*)
- and *padding_cells_by_level* is how many layers of the current level of cells before cell size in increased

For *padding_cells_by_level*, I would personally like the have the following options:
- *int*: where the same number of padding cells is provided for every level
- *array_like* of *int*: since discretization below *level* is not done, we don't want define values in this argument that aren't used. If it is an array_like, the first entry provides the padding for refinement at *level* and we go from.

Reply:
Hmm, that's also an interesting way of looking at describing the padding, and would disconnect the `level` and `padding_cells_by_level` argument. I'll ponder on that a little bit.